File: design_patterns_101/Additional Patterns/Iterator Pattern/iteratorPattern.js
// Concrete Iterator
class AlphabeticalOrderIterator {
    constructor(collection, reverse = false) {
      this.collection = collection;
      this.reverse = reverse;
      this.position = reverse ? collection.length - 1 : 0;
    }
  
    next() {
      const result = this.collection[this.position];
      this.position += this.reverse ? -1 : 1;
      return result;
    }
  
    hasNext() {
      if (this.reverse) {
        return this.position >= 0;
      }
      return this.position < this.collection.length;
    }
  }
  
  // Aggregate interface
  class IterableCollection {
    createIterator() {
      throw new Error("Abstract method!");
    }
  }
  
  // Concrete Aggregate
  class WordsCollection extends IterableCollection {
    constructor() {
      super();
      this.collection = [];
    }
  
    add(item) {
      this.collection.push(item);
    }
  
    get(index) {
      return this.collection[index];
    }
  
    createIterator() {
      return new AlphabeticalOrderIterator(this.collection);
    }
  
    reverseIterator() {
      return new AlphabeticalOrderIterator(this.collection, true);
    }
  }
  
  // Usage
  const collection = new WordsCollection();
  collection.add("First");
  collection.add("Second");
  collection.add("Third");
  
  console.log("Straight traversal:");
  const iterator = collection.createIterator();
  while (iterator.hasNext()) {
    console.log(iterator.next());
  }
  
  console.log("\nReverse traversal:");
  const reverseIterator = collection.reverseIterator();
  while (reverseIterator.hasNext()) {
    console.log(reverseIterator.next());
  }
  

File: design_patterns_101/Additional Patterns/Iterator Pattern/iteratorPattern.py
from __future__ import annotations
from collections.abc import Iterable, Iterator
from typing import Any, List

# Concrete Iterator
class AlphabeticalOrderIterator(Iterator):
    _position: int = None
    _reverse: bool = False

    def __init__(self, collection: WordsCollection, reverse: bool = False) -> None:
        self._collection = collection
        self._reverse = reverse
        self._position = -1 if reverse else 0

    def __next__(self):
        try:
            value = self._collection[self._position]
            self._position += -1 if self._reverse else 1
        except IndexError:
            raise StopIteration()

        return value

# Aggregate interface
class IterableCollection(Iterable):
    @abstractmethod
    def create_iterator(self) -> Iterator:
        pass

# Concrete Aggregate
class WordsCollection(IterableCollection):
    def __init__(self, collection: List[Any] = []) -> None:
        self._collection = collection

    def __iter__(self) -> AlphabeticalOrderIterator:
        return AlphabeticalOrderIterator(self)

    def create_iterator(self) -> AlphabeticalOrderIterator:
        return AlphabeticalOrderIterator(self)

    def add_item(self, item: Any) -> None:
        self._collection.append(item)

    def __getitem__(self, index: int) -> Any:
        return self._collection[index]

# Usage
collection = WordsCollection()
collection.add_item("First")
collection.add_item("Second")
collection.add_item("Third")

print("Straight traversal:")
for element in collection:
    print(element)

print("\nReverse traversal:")
iterator = collection.create_iterator()
while iterator:
    try:
        print(iterator.__next__())
    except StopIteration:
        break
